Dynamic namespace support

thallgren opened the pull request:
This PR changes how the Telepresence traffic-manager and mutating webhook deals with namespaces so that everything is controlled by one single namespace selector.

The top level Helm chart value `namespaceSelector` was added. It is a verbatim Kubernetes namespace selector, and deprecates the `agentInjector.webhook.namespaceSelector` and `managerRbac.namespaces`. It also acts as the default for `clientRbac.namespaces`.

The short form value `namespaces` was also added. It is a list of names and mutually exclusive to `namespaceSelector`. The list will get transformed into:
```yaml
namespaceSelector:
  matchExpressions:
  - key: kubernetes.io/metadata.name
    operator: In
    values: [ <namespaces> ]
```

A selector can be _dynamic_ or _static_. This in turn controls if telepresence is "cluster-wide" or "namespaced". A _dynamic_ selector
requires cluster-wide permissions for the traffic-manager, and only a _static_ selector can serve as base when installing `Role`/`RoleBinding` pairs.

A selector is considered static if it meets the following conditions:
- The selector must have exactly one element in the `matchLabels` or the
  `matchExpression` list (if the element is in the `matchLabels` list,
   it is normalized into "key in [value]").
- The element must meet the following criteria:
  - The `key` of the element must be "kubernetes.io/metadata.name".
  - The `operator` of the element must be "In" (case-sensitive).
  - The `values` list of the element must contain at least one value.

thallgren committed:
Dynamic namespace support.

This commit changes how the Telepresence traffic-manager and mutating
webhook deals with namespaces so that everything is controlled by one
single namespace selector.

A selector can be _dynamic_ or _static_. This in turn controls if
telepresence is "cluster-wide" or "namespaced". A _dynamic_ selector
requires cluster-wide permissions for the traffic-manager, and only a
_static_ selector can serve as base when installing Role/RoleBinding
pairs.

A selector is considered static if it meets the following conditions:
- The selector must not have `matchLabels`.
- The selector must have exactly one element in the `matchLabels` or the
  `matchExpression` list (if the element is in the `matchLabels` list,
   it is normalized into "key in [value]").
- The element must meet the following criteria:
  - The `key` of the element must be "kubernetes.io/metadata.name".
  - The `operator` of the element must be "In" (case-sensitive).
  - The `values` list of the element must contain at least one value.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Sort list of Helm values alphabetically.

The current order is very incosistent and hard to follow. This commit
sorts the list alphabetically, without exceptions.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update the Helm chart README.

Add description of new `namespaces` and `namespaceSelector` values, and
deprecate the `namespaced` and `namespaces` in `client.Rbac` and
`manager.Rbac`.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update documentation with respect to namespace selection.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update CHANGELOG.md with respect to namespace selection.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Integration tests for namespace selectors.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix incorrect namespaceSelector assignment in integration tests.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that podSubnetWatcher includes the traffic-manager podIP.

The traffic-manager's IP is used for DNS in the client, so it must be
included in when computing the pod subnets. It must be added explicitly
now, because it's no longer required that the traffic-manager's own
namespace is managed.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip commented on charts/telepresence/templates/trafficManagerRbac/cluster-scope.yaml:
```diff
@@ -60,7 +54,7 @@ rules:
   - get
   - watch
-{{- if .Values.agentInjector.enabled }}
+{{- if .agentInjector.enabled }}
```
I know it's an older code, but I believe this could be merged with the above. We'd only need to add `create` here and the previous definition could be dropped.

thallgren replied:
No, I don't think that's correct. This entry is specifically for the configmap named "telepresence-agents".  The entry above isn't (using create isn't possible for a named resource).

P0lip commented on cmd/traffic/cmd/manager/cluster/info.go:
```diff
@@ -304,6 +296,21 @@ func clusterDomainFromResolvConf(confFile, namespace string) (string, error) {
 }
 
 func (oi *info) watchNodeSubnets(ctx context.Context, mustSucceed bool) bool {
+	ok, err := k8sapi.CanI(ctx, &auth.ResourceAttributes{
```
A loose suggestion, but it'd be nice if `k8sapi.CanI` took an slice of `*auth.ResourceAttributes` or (make it a variadic function), so we could dedupe some of the code here and one or two other places.

P0lip commented on cmd/traffic/cmd/manager/config/config.go:
```diff
@@ -1,60 +1,81 @@
 }
 
+func (ae AgentEnv) Equal(o AgentEnv) bool {
+	return slices.Equal(ae.Excluded, o.Excluded)
```
a minor nit, but the order of elements shouldn't really matter in this case, so we could potentially account for it like we do in some other places.
We don't seem to sort them anywhere, we could do that after the data is unmarshaled.

P0lip commented on cmd/traffic/cmd/manager/namespaces/watcher.go:
```diff
@@ -0,0 +1,297 @@
+		}
+		if err != nil {
+			return false, fmt.Errorf("error listing namespaces: %v", err)
+		}
```
```suggestion
	  return false, fmt.Errorf("error listing namespaces: %v", err)
```

P0lip reviewed (approved):
LGTM. I like this change.

thallgren committed:
Update cmd/traffic/cmd/manager/namespaces/watcher.go

Co-authored-by: Jakub Rożek <jrozek@datawire.io>
Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Changes in response to code-review.

- Make k8sapi.CanI a variadic function (and use where applicable).
- Sort AgentEnv.Excluded before checking slices for equality.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix bad indentation.

Signed-off-by: Thomas Hallgren <thomas@tada.se>
